keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,134 @@
+---
+title: declarative-field-manager
+authors:
+  - "@kwiesmueller"
+owning-sig: sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-03-24
+last-updated: yyyy-mm-dd
+status: implemented
+---
+
+# declarative-field-manager
+
+## Summary
+
+Enable users to set the fieldManager, introduced with Server Side Apply, declaratively when sending the object to the apiserver.
+The Server Side Apply feature introduced ownership of fields to improve merging and improve the object lifecycle when multiple actors interact with the same object.
+
+While currently the identifier of the current actor (fieldManager) is set through an request option or the request user-agent,
+this KEP aims to provide a declarative way of setting the current fieldManager through a field in the object itself.
+
+## Motivation
+
+It is a common practice to use Kubernetes manifests as stored configuration of an application. Either to just introduce version control to the cluster's state, or to automate the process of updating objects.
+
+With the addition of Server Side Apply it could become necessary to change the way those manifests get applied, as the most benefit from this new feature requires every actor to define it's fieldManager. Doing this, currently is only possible through api options, the user-agent or as a kubectl option.
+
+To keep the way of interacting with manifests as easy as possible, this KEP suggests to allow defining the fieldManager as part of the manifest itself. Therefor no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.
+
+This would also assist existing solutions already interacting with manifest files, as the fieldManager information gets provided in a way they already know to handle which might improve compatibility.
+
+### Goals
+
+- Allow users to set their fieldManager information declaratively in their manifests
+
+### Non-Goals
+
+- Replace the existing ways to set the fieldManager information
+
+## Proposal
+
+To realize the goals, we would add a field to ObjectMeta called `fieldManager`, right next to the existing `managedFields`.
```

Comment:
We should figure out a name that is very clear. E.g. "field" isn't helpful because the manager is for the entire object, I think.

Maybe we can consider putting this in a `WriteOnly` or `DeclarativeWriteOnlyQueryParameters` sub-struct? And just call it Manager? I'm sure someone can think of an improvement to that.

Reply:
`DeclarativeWriteOnlyQueryParameters` reads a little bit like `annotations` in that case

From [kubernetes annotations documentation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/):
> Here are some examples of information that could be recorded in annotations:
> - Fields managed by a declarative configuration layer. Attaching these fields as annotations distinguishes them from default values set by clients or servers, and from auto-generated fields and fields set by auto-sizing or auto-scaling systems.

Would you be open to `fieldsManager`?

Reply:
If you place it in an annotation, you absolutely have to have kubectl delete the annotation before writing/applying--otherwise, appliers will fight over it. This is not data a declarative config layer manages about an *object*. No matter how we do it, it should not be stored.

"fieldsManager" is not my first choice, at least not if loose in metadata. Here's another idea?

```
...
metadata:
  applyArguments:
    - manager: my-config-system
```

Reply:
Not excited about the extra-depth. `applyManager`

Reply:
What other potential apply args could we have to make it it's own object?
The first draft was just `manager` btw.
Another one could be `applyManager` or `applier`

Reply:
Oh, too slow with applyManager :-)

Reply:
Alternative option: `metadata.applyOptions.manager`.

I think I'm still reluctant to adding an extra layer of nesting since:
1. This is really something people will have to type in their config, I wish it was simpler,
2. I don't see a great argument for the added complexity: we don't foresee any other additional "arguments"
3. Documentation can probably fill the gap for people to understand that this is a write-only field. 

Alternatively we can use maybe `applyManager(Option|Argument)` to clarify further.

My favorite option still is `applyManager` though, I could settle for `applyManagerOption` and would rather avoid the nesting.

Reply:
Thanks @apelisse!
Another idea I had when starting the KEP was putting information like that next to or into TypeMeta.

So as an example:
```
apiVersion: v1
kind: ...
options:
  applyManager: "test"
metadata:
...
```

I know this change is a little radical and has a bigger influence on the general api.
After some talking about it I/we opted for putting it into metadata, but it would be an option I thought about where the field might be more in place, as it never will be part of the object anyways (write-only).
And it would fit from a user perspective as apiVersion and kind already are special regarding this.

Reply:
That's definitely radical, but it reads very nice from a user's perspective. @liggitt do you think we can sell this idea (or a variant) to folks?

Reply:
I think it's too late to claim ownership of another top-level field... the potential for conflicts with existing objects is too great, and keeping the special knowledge and behavior bounded to the metadata struct is helpful for custom resource schemas and validation

Reply:
What about ... `metadata.options.applyManager`?

Reply:
We'll, I guess we need some middleway.
What do you think @lavalamp ?